compiler/src/iree/compiler/Codegen/Common/TransformDialectStrategies.cpp

```diff
@@ -0,0 +1,614 @@
+///      fusion, no interleaved canonicalization/cse/enabling occur and the
+///      resulting fusion may not be as good.
+///   2. In iterative mode, the user is responsible for providing the fusion
+///      order but interleaved canonicalization/cse/enabling occur which may
+///      result in better fusion results.
+///      TODO: Transform dialect op to apply topological sort and avoid the
+///      manual intervention.
+/// If `resultingFusedOpsHandles` is a non-null pointer, the fused operation are
+/// appended in order.
+// TODO: apply forwarding pattern.
+template <typename TilingTransformOp, typename TileOrNumThreadSpec>
+static Value tileAndFuseAndDistributeImpl(
+    ImplicitLocOpBuilder &b, Value rootH, ValueRange opsHToFuse,
+    ArrayRef<OpFoldResult> tileSizesOrNumThreads, ArrayAttr threadDimMapping,
+    TileFuseDistOpt options, SmallVectorImpl<Value> *resultingFusedOpsHandles) {
+  auto tileToForeachOp = b.create<TilingTransformOp>(
+      rootH, tileSizesOrNumThreads, TileOrNumThreadSpec(), threadDimMapping);
+  Value foreachThreadH = tileToForeachOp.getForeachThreadOp();
+  if (options.batchFusion) {
+    Value mergedOpsH =
+        b.create<MergeHandlesOp>(opsHToFuse, /*deduplicate=*/true);
+    b.create<FuseIntoContainingOp>(mergedOpsH, foreachThreadH);
+    assert(!resultingFusedOpsHandles && "Handle needs unpacking");
+  } else {
+    for (Value h : opsHToFuse) {
+      Value fusedH = b.create<FuseIntoContainingOp>(h, foreachThreadH);
+      if (resultingFusedOpsHandles) resultingFusedOpsHandles->push_back(fusedH);
+    }
+  }
+  return foreachThreadH;
+}
+
+/// Call tileAndFuseAndDistributeImpl with ArrayRef<int64_t> tilesSizes.
+template <typename TilingTransformOp = TileToForeachThreadOp>
+static Value buildTileFuseDistWithTileSizes(
+    ImplicitLocOpBuilder &b, Value rootH, ValueRange opsHToFuse,
+    ArrayRef<int64_t> tileSizes, ArrayAttr threadDimMapping,
+    TileFuseDistOpt options = TileFuseDistOpt(),
+    SmallVectorImpl<Value> *resultingFusedOpsHandles = nullptr) {
+  return tileAndFuseAndDistributeImpl<TilingTransformOp,
+                                      transform::TileSizesSpec>(
+      b, rootH, opsHToFuse, getAsOpFoldResult(b.getI64ArrayAttr(tileSizes)),
+      threadDimMapping, options, resultingFusedOpsHandles);
+}
+
+/// Call tileAndFuseAndDistributeImpl with a handle to multiple tilesSizes.
+template <typename TilingTransformOp = TileToForeachThreadOp>
+static Value buildTileFuseDistWithTileSizes(
+    ImplicitLocOpBuilder &b, Value rootH, ValueRange opsHToFuse,
+    Value tileSizeHandle, ArrayAttr threadDimMapping,
+    TileFuseDistOpt options = TileFuseDistOpt(),
+    SmallVectorImpl<Value> *resultingFusedOpsHandles = nullptr) {
+  return tileAndFuseAndDistributeImpl<TilingTransformOp,
+                                      transform::TileSizesSpec>(
+      b, rootH, opsHToFuse, ArrayRef<OpFoldResult>{tileSizeHandle},
+      threadDimMapping, options, resultingFusedOpsHandles);
+}
```

Comment:
There are 4 parameters controlling the transformation, the suggestion is to splice one of them out into a struct for the purpose of removing one function call.

This is a bad tradeoff because distributing the control logic between 2 mechanisms (parameters and an options struct), loses uniformity and increases API confusion: we should use the same mechanism for everything.

Putting everything into a single options struct would be an anti-pattern for an MLIR build-like API and we should not be inventive here.

In fact, thinking about this deeper, I should have pushed back against the options struct to control fusion in the first place, the traditional MLIR solution here is more builders and more parameters.

> (It's always good to reduce the number of function arguments if there are many function arguments and the order matters, see https://abseil.io/tips/173)

It's really not when it goes against established practices and uniformity with the code around you. 
The MLIR/LLVM coding standard is what we should follow.
There may be a case to be made here to the MLIR community as to how to improve these things globally and uniformly.
This will involve deeper Tablegen surgery so I am not going to hold my breath on that :)